apache/zeppelin
# [Zeppelin 1165 : improve]  Paragraph dynamic relationship executable form

### What is this PR for?

Workflow process feature.
(To ensure the success of each paragraph, it is possible to run consecutively.)
#### Case 1

Through a dynamic form, you can execute the order in paragraph.
There is a difference with traditional methods.
Please check the following flowchart.
![workflowdynamicformcontrol](https://cloud.githubusercontent.com/assets/10525473/16791726/a4b96ff0-48fc-11e6-8e23-9ec577066bb7.png)
#### Case 2

In general, when run a plurality of Paragraph, it performs Note entire run.
This is a good way to run a lot of Paragraph contained in the Note.
However, the problem occurs if the Interpreter of Paragraph different.
![notebook_example](https://cloud.githubusercontent.com/assets/10525473/16803203/175ad01a-4940-11e6-8949-72d0c49bdf9e.png)
For Paragraph each using a different type of one of the Interpreter Note but run in sequence, the end is all different.

![normal notebook run](https://cloud.githubusercontent.com/assets/10525473/16803193/069c4d94-4940-11e6-9293-888b6c6288a0.png)
For example, Markdown is a very fast Interpreter.
The process is completed very quickly.
This is a problem in the sequential execution Paragraph.

![worklfow run](https://cloud.githubusercontent.com/assets/10525473/16803192/06998122-4940-11e6-8f01-43cdf64f2eef.png)
This feature ensures a certain execution order Notebook with each Interpreter.
##### Case 3

For concurrent job in the workflow ...

![job_repl](https://cloud.githubusercontent.com/assets/10525473/16828860/96e601b8-49ce-11e6-87d0-6f7fc30ce751.png)

If the current functional design is supposed to run at the same time, as follows 
It is to share the results of the job.
But if the situation need to run the job at the same time, subject to their execution flow.

*\* The results will have to succeed, the following paragraph will be executed. **
### What type of PR is it?

Improvement
### jira

https://issues.apache.org/jira/browse/ZEPPELIN-1165
### How should this be tested?

Dynamic form
For usage, see the following: 
https://zeppelin.apache.org/docs/0.7.0-SNAPSHOT/manual/dynamicform.html

In the markdown type:

```
%md
${workflow:workflow=MyWorkflowName,NotebookId01:ParagraphId01|NotebookId02:ParagraphId02}
```
- details explain.

You can specify the title work flow.
Simply displayed on the UI. 

```
${workflow:workflow=
or
${workflow:workflow=MyWorkflow
```

title name : MyWorkflow

character a comma(,), you can start working set.

```
${workflow:workflow=MyWorkflow**,**
```

Enter your Notebook ID and Paragraph ID.

```
${workflow:workflow=MyWorkflow,2BS99R6QA:20160712-163556_1475991024}
```

In order to distinguish between the Notebook ID and Paragraph ID, characters <:> it is.
The example was set to run 20160712-163556_1475991024 paragraph with a notebook id called 2BS99R6QA.
- how to get notebook id?
  <img width="534" alt="getnotebookid" src="https://cloud.githubusercontent.com/assets/10525473/16795411/30faab1c-4918-11e6-8ef3-8b6fc6f6f039.png">
- how to get paragraph id?
  <img width="810" alt="getparagraphid" src="https://cloud.githubusercontent.com/assets/10525473/16795412/31041d8c-4918-11e6-9b9f-8f472692e3e1.png">

Through a pipe character( | ), you can add tasks consecutively.

```
${workflow:workflow=MyWorkflow,2BS99R6QA:20160712-163556_1475991024|2BS99R6QA:20160710-03221_12234323}
```
#### Note run

![cap 2016-07-14 15-11-07-036](https://cloud.githubusercontent.com/assets/10525473/16829747/732164a0-49d5-11e6-938f-9c08cfb772c1.png)

If you want to run to all Paragraph in  Note, you can:

```
NoteID : *
or
NoteID
```

```
${workflow:workflow=MyworkflowName,NotebookId}
or
${workflow:workflow=MyworkflowName,NotebookId:*}
```

for example

```
${workflow:workflow=Myworkflow,2A94M5J1Z}
```

Of course, it is possible to run a particular paragraph and mix.

```
%md
${workflow:workflow=Myworkflow,2A94M5J1Z|r|2BQWXDH1C:20160714-151024_1284650101}
```
### Screenshots (if appropriate)

![workflow](https://cloud.githubusercontent.com/assets/10525473/16830111/d54e909c-49d7-11e6-87c2-0de4a3b5b793.gif)
### Questions:
- Does the licenses files need update? no
- Is there breaking changes for older versions? no
- Does this needs documentation? yes, After the merge.
